vivo-project/Vitro
# Browse filter results on page

[VIVO PR](https://github.com/vivo-project/VIVO/pull/4005)
# What does this pull request do?

- new page content type "Browse search filter results" to display search results limited by selected filter. Also additional filters and sort options could be defined in custom page template (template [example](https://github.com/TIBHannover/tib-fis-installer/blob/1.15.1-snapshot/project-installer/webapp/src/main/webapp/templates/freemarker/body/menupage/browsePersonSearchFilterValues.ftl))
- new facet autocomplete input field that appears if number of facets in bigger than more limit defined for the filter
- new property to limit variables of document modifiers to be populated into search indexes

# What's new?
Refactored PagedSearchController. 
Added search filter option to ManagePageGenerator
Refactored ProcessN3GetterDataMap
Created new data getter class display:SearchFilterValuesDataGetter , java implementation in SearchFilterValuesDataGetter.java
Created object property search:direction to configure sort direction of filter values and search results.
Created filters to filter result in alphabetical indexes 
:filter_label_regex - default regex filter that uses i18n field :field_label_display with name locale + _label_display
:filter_raw_label_regex - fallback regex filter that uses nameRaw search field for VIVO instances without i18n support
search:limitDisplayTo object property was created to limit user access to display of search groups, filters, values.
Resolved bug on search page which appears if user select too many filters. In that case search results are empty and selected filter values are not shown on the page, so the user can't unselect them. 
Added options to sort results, number of results on page and "Download results" options on custom pages

Applied changes for search ontology (from [ontology interests group meeting](https://wiki.lyrasis.org/display/VIVO/2025-01-08+Ontology+Interest+Group+Call))
Removed search:isAscending boolean data property,  new object property search:direction should be used instead. 
(search:isAscending is still could be used for backwards compatibility).
Removed search:reverseFacetOrder boolean data property, new object property search:direction should be used instead. 
(search:reverseFacetOrder is still could be used for backwards compatibility).
Renamed search:order integer data property into search:rank, 
(search:order is still could be used for backwards compatibility).
Created class search:SortDirection and two individuals search-ind:ascending and search-ind:descending to be used with search:direction object property.
Added facet autocomplete input field that is being visible in case filter has number of values greater than limit of facets to be shown before "more..." button

# How should this be tested?
How to test new page content type:
* Create a new page in Page management menu. 
* For the content select "Browse search filter results", select a type and save it.
* To select custom template select Template option "Custom template requiring content" and enter filename of custom Freemarker template in input field (Optional).
* Open page, try selecting facets, alphabetical index, pagination
* Test above in VIVO with and without language filtering

How to test facet autocomplete input
* To test facet autocomplete input you would need a filter that doesn't use URIs to store in search index. [An example](https://github.com/litvinovg/VIVO/compare/browse-filter-results...litvinovg:VIVO:facet-autocomplete-test-settings) of modification of organization filter
* Run full search re-index after filter modification.
* You can check how it work on /search page by selecting the filter and entering at least 3 characters in input field. 
* Suggestions should appear. Select a suggestion and page should reload with applied selected facet.
 
# Additional Notes:
Examples of customized [templates](https://github.com/TIBHannover/tib-fis-installer/tree/1.15.1-snapshot/project-installer/webapp/src/main/webapp/templates/freemarker/body/menupage), [document modifiers](https://github.com/TIBHannover/tib-fis-installer/blob/1.15.1-snapshot/project-installer/home/src/main/resources/rdf/display/everytime/searchFieldsConfigurationVivo.n3)
This change require documentation to be updated.

# Interested parties
@hauschke @VIVO-project/vivo-committers

# Reviewers' expertise

Candidates for reviewing this PR should have some of the following expertises:
1. Java
2. JavaScript
3. FreeMarker
4. SPARQL
5. Ontologies
1. Natural language knowledge
    1. English
    2. German
    3. Spanish
    4. French
    5. Portuguese
    6. Russian
    7. Serbian

# Reviewers' report template
**Please update the following template which should be used by reviewers.**
## General comment
A reviewer should provide here comments and suggestions for requested changes if any.
## Testing
A reviewer should briefly describe here how it was tested
## Code reviewing
A reviewer should briefly describe here which part was code reviewed
